### PR TITLE
Add junit5 support - fixes #489

### LINKED ISSUE #489: JUnit 5 support
Currently Spring cloud contract supports JUnit 4 and Spock. It would be nice to have Junit 5 too

### PR DESCRIPTION
**Pull request by OlgaMaciaszek**
Fixes https://github.com/spring-cloud/spring-cloud-contract/issues/489

### CODE REVIEW

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/MethodBuilder.groovy [hunk 1]
```diff
@@ -104,26 +108,29 @@ class MethodBuilder {
...
 			return new MockMvcSpockMethodRequestProcessingBodyBuilder(stubContent, configProperties)
-		} else if (configProperties.targetFramework == TestFramework.SPOCK) {
+		} else if (configProperties.targetFramework == SPOCK) {
 			return new MockMvcSpockMethodRequestProcessingBodyBuilder(stubContent, configProperties)
 		}
 		return new MockMvcJUnitMethodBodyBuilder(stubContent, configProperties)
 	}
 
+	private boolean isJUnitType() {
+		JUNIT == configProperties.targetFramework || JUNIT5 == configProperties.targetFramework
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/MethodBuilder.groovy]: `return`

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/BaseImportProvider.groovy [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import static org.springframework.cloud.contract.verifier.config.TestFramework.CUSTOM
+import static org.springframework.cloud.contract.verifier.config.TestFramework.JUNIT
+import static org.springframework.cloud.contract.verifier.config.TestFramework.JUNIT5
+import static org.springframework.cloud.contract.verifier.config.TestFramework.SPOCK
+
+
+/**
+ * Provides imports based on test framework.
+ *
+ * @author Olga Maciaszek-Sharma
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/BaseImportProvider.groovy]: `@since 2.1.0`

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/BaseImportProvider.groovy [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+	 * Returns list of static imports for provided test framework.
+	 * @param testFramework
+	 * @return
+	 */
+	static List<String> getStaticImports(TestFramework testFramework) {
+		return GENERAL_IMPORTS.staticImports + TEST_FRAMEWORK_SPECIFIC_IMPORTS.get(testFramework).staticImports
+	}
+
+	/**
+	 * Returns jUnit rule import.
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/BaseImportProvider.groovy]: Let's remove that javadoc - it's self-explanatory

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/HttpImportProvider.groovy [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import static org.springframework.cloud.contract.verifier.config.TestFramework.JUNIT5
+import static org.springframework.cloud.contract.verifier.config.TestFramework.SPOCK
+import static org.springframework.cloud.contract.verifier.config.TestMode.EXPLICIT
+import static org.springframework.cloud.contract.verifier.config.TestMode.JAXRSCLIENT
+import static org.springframework.cloud.contract.verifier.config.TestMode.MOCKMVC
+
+/**
+ * Provides imports based on test framework and test mode.
+ *
+ * @author Olga Maciaszek-Sharma
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/HttpImportProvider.groovy]: `@since`

Comment (marcingrzejszczak) [reply spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/HttpImportProvider.groovy]: `@CompileStatic`

Comment (marcingrzejszczak) [reply spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/HttpImportProvider.groovy]: `@PackageScope` ?

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/BaseImportProvider.groovy (hunk 2 above)
Comment (marcingrzejszczak) [reply spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/BaseImportProvider.groovy]: `@PackageScope` ?

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/ImportDefinitions.groovy [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package org.springframework.cloud.contract.verifier.builder.imports
+
+/**
+ * @author Olga Maciaszek-Sharma
+ */
+
+class ImportDefinitions {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/ImportDefinitions.groovy]: `@since`, `@CompileStatic`, `@PackageScope` - I'm not sure about package scopes since people might want to reuse these classes if they want their own java tests

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/MessagingImportProvider.groovy [hunk 6]
```diff
@@ -0,0 +1,29 @@
+package org.springframework.cloud.contract.verifier.builder.imports
+
+/**
+ * Provides imports based on test framework and test mode.
+ *
+ * @author Olga Maciaszek-Sharma
+ */
+class MessagingImportProvider {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/MessagingImportProvider.groovy]: `@since`, `@CompileStatic`

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/MessagingImportProvider.groovy [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+	/**
+	 * Returns list of imports for messaging test contracts.
+	 * @return list of imports
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/MessagingImportProvider.groovy]: Either first or last line of the javadoc

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/TestFramework.groovy [hunk 8]
```diff
@@ -16,68 +16,105 @@
...
+	 * @param classModifier
+	 * @param methodModifier
+	 * @param lineSuffix
+	 * @param classExtension
+	 * @param classNameSuffix
+	 * @param ignoreClass
+	 * @param orderAnnotationImports
+	 * @param orderAnnotation
+	 */
+	@Deprecated
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/TestFramework.groovy]: 👍

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/CustomDefinition.groovy [hunk 9]
```diff
@@ -0,0 +1,64 @@
+package org.springframework.cloud.contract.verifier.config.framework
+
+/**
+ * Defines elements characteristic of Custom test framework to be used during test class construction.
+ *
+ * @author Olga Maciaszek-Sharma
+ */
+class CustomDefinition implements TestFrameworkDefinition {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/CustomDefinition.groovy]: `@since`, `@CompileStatic`

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/JUnit5Definition.groovy [hunk 10]
```diff
@@ -0,0 +1,64 @@
+package org.springframework.cloud.contract.verifier.config.framework
+
+/**
+ * Defines elements characteristic of JUnit5 test framework to be used during test class construction.
+ *
+ * @author Olga Maciaszek-Sharma
+ */
+class JUnit5Definition implements TestFrameworkDefinition {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/JUnit5Definition.groovy]: `@since`, `@CompileStatic`

File: docs/src/main/asciidoc/verifier_setup.adoc [hunk 11]
```diff
@@ -245,8 +245,8 @@ from the Groovy DSL should be placed. By default its value is
 `$buildDir/generated-test-sources/contractVerifier`.
 * *stubsOutputDir*: Specifies the directory where the generated WireMock stubs from
 the Groovy DSL should be placed.
-* *targetFramework*: Specifies the target test framework to be used. Currently, Spock and
-JUnit are supported with JUnit being the default framework.
+* *targetFramework*: Specifies the target test framework to be used. Currently, Spock, JUnit 4 (`TestFramework.JUNIT` and
+JUnit 5 are supported with JUnit 5 being the default framework.
```
Comment (marcingrzejszczak) [inline docs/src/main/asciidoc/verifier_setup.adoc]: `JUnit 4` should be the default framework IMO

File: docs/src/main/asciidoc/verifier_setup.adoc [hunk 12]
```diff
@@ -582,8 +582,8 @@ classes.
 use Spock classes, the class is `spock.lang.Specification`.
 * *contractsDirectory*: Specifies a directory containing contracts written with the
 GroovyDSL. The default directory is `/src/test/resources/contracts`.
-* *testFramework*: Specifies the target test framework to be used. Currently, Spock and
-JUnit are supported with JUnit being the default framework
+* *testFramework*: Specifies the target test framework to be used. Currently, Spock, JUnit 4 (`TestFramework.JUNIT` and
+JUnit 5 are supported with JUnit 5 being the default framework.
```
Comment (marcingrzejszczak) [inline docs/src/main/asciidoc/verifier_setup.adoc]: `JUnit 4` should be the default framework IMO

File: spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/src/main/java/org/springframework/cloud/contract/maven/verifier/GenerateTestsMojo.java [hunk 13]
```diff
@@ -21,6 +21,7 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/src/main/java/org/springframework/cloud/contract/maven/verifier/GenerateTestsMojo.java]: Can we revert these changes?

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/JUnitDefinition.groovy [hunk 14]
```diff
@@ -0,0 +1,64 @@
+package org.springframework.cloud.contract.verifier.config.framework
+
+/**
+ * Defines elements characteristic of JUnit test framework to be used during test class construction.
+ *
+ * @author Olga Maciaszek-Sharma
+ */
+class JUnitDefinition implements TestFrameworkDefinition {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/JUnitDefinition.groovy]: `@since`, `@CompileStatic`

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/SpockDefinition.groovy [hunk 15]
```diff
@@ -0,0 +1,64 @@
+package org.springframework.cloud.contract.verifier.config.framework
+
+/**
+ * Defines elements characteristic of Spock test framework to be used during test class construction.
+ *
+ * @author Olga Maciaszek-Sharma
+ */
+class SpockDefinition implements TestFrameworkDefinition {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/SpockDefinition.groovy]: `@since`, `@CompileStatic`

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/TestFrameworkDefinition.groovy [hunk 16]
```diff
@@ -0,0 +1,31 @@
+package org.springframework.cloud.contract.verifier.config.framework
+
+/**
+ * Defines elements characteristic of a given test framework to be used during test class construction.
+ *
+ * @author Olga Maciaszek-Sharma
+ */
+interface TestFrameworkDefinition {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/config/framework/TestFrameworkDefinition.groovy]: `@since`, `@CompileStatic` + javadocs on methods with an example. 

E.g. (I don't know if that's correct with `{@code ;}`)

```groovy
/**
* Returns the characters that should end each line. E.g. for Java tests that would mean {@code ;}
**/
String getLineSuffix() 
```

File: spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/src/main/java/org/springframework/cloud/contract/maven/verifier/GenerateTestsMojo.java (hunk 13 above)
Comment (OlgaMaciaszek) [reply spring-cloud-contract-tools/spring-cloud-contract-maven-plugin/src/main/java/org/springframework/cloud/contract/maven/verifier/GenerateTestsMojo.java]: Sure, but this was the effect of formatting it with the official Spring project IDE formatter.

**Comment by OlgaMaciaszek** [comment]
Thanks :) . I don't think we need more tests, since I have added JUNIT5 to all test scenarios that were earlier added for the previously existing test modes. Will fix the rest till tomorrow.

File: spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/HttpImportProvider.groovy (hunk 4 above)
Comment (OlgaMaciaszek) [reply spring-cloud-contract-verifier/src/main/groovy/org/springframework/cloud/contract/verifier/builder/imports/HttpImportProvider.groovy]: I would leave out `@CompileStatic`. Would have to cast between String and GString for `ImportDefinitions` constructor.
